File: prolog/story/police.pl
:- multifile holding/1.

% Info

info(police_station) :-
    write('You are inside of a police building that has a lot of traffic is bustling with activity,'), nl,
    write('with officers and visitors moving quickly through the halls.'), nl,
    write('The sound of phones ringing'), nl,
    write('and radios crackling can be heard constantly in the background.'), nl,
    write('The walls are lined with bulletin boards'), nl,
    write('and posters filled with information and updates related to ongoing cases.').

info(policeman_office) :-
    write('You are in office of a policeman who is very busy is cluttered'), nl,
    write('with piles of paperwork and files,'), nl,
    write('as well as various items of police equipment scattered throughout the room.'), nl,
    write('The desk is usually covered with a computer,'), nl,
    write('radio, phone, and other tools of the trade,'), nl,
    write('with little space for anything else.').

info(cage) :-
    write('You see cells at the police station'), nl,
    write('where there are many detainees are small and cramped,'), nl,
    write('with just enough room'), nl,
    write('for a narrow bed and a toilet in the corner.'), nl,
    write('The walls are made of cold, gray concrete'), nl,
    write('and the only light comes from a small window high up near the ceiling.').

% People

person_at(scary_prisoner, cage).
subject(scary_prisoner, reason, 'I killed a man').

person_at(policeman, police_station).
subject(policeman, help, 'I cannot help you. I''m busy at the moment!').

person_at(sheriff, policeman_office).
subject(sheriff, case, 'We are looking for a #1 wanted interpol criminal.\nHe was seen fleeing into the forest!').
subject(sheriff, employees_criminal_record, 'TODO: jestesmy zajeci, wyslemy poczta').

conversation_result(sheriff, employees_criminal_record) :- 
    retractall(asked_for_criminal_record), 
    assert(asked_for_criminal_record).

% Items

at(makeshift_knife, cage).
info(makeshift_knife) :- 
    write('Better take it from here!'), nl,
    write('I don''t want it to fall into the wrong hands'), nl.

at(weed, cage) :- holding(four_leaf_clover).

at(handcuffs, police_station).
at(gun, police_station).
at(ammunition, police_station).
at(suitcase, police_station).

at(pen, policeman_office).
at(paper, policeman_office).
at(blank_paper, policeman_office).

File: haskell/src/Story/Police.hs
module Story.Police where

import Rules.Item
import Rules.Location
import Rules.Person

-- Items

makeshiftKnife =
  Item
    "makeshift_knife"
    [ "Better take it from here!",
      "I don''t want it to fall into the wrong hands"
    ]
    Nothing

-- People

scaryPrisoner =
  Person
    "scary_prisoner"
    [Subject "reason" "I killed a man"]

policeman =
  Person
    "policeman"
    [Subject "help" "I cannot help you. I'm busy at the moment!"]

-- TODO: add conversation result
sheriff =
  Person
    "sheriff"
    [ Subject "case" "We are looking for a #1 wanted interpol criminal.\nHe was seen fleeing into the forest!",
      Subject "employees_criminal_record" "What are you here for? Can''t you see we don''t have time now? We''ll mail everything to you, now get out, because as you can see, we''re very busy."
    ]

-- Locations

policeStation =
  Location
    "Police Station"
    [ "You are inside of a police building that has a lot of traffic is bustling with activity,",
      "with officers and visitors moving quickly through the halls.",
      "The sound of phones ringing",
      "and radios crackling can be heard constantly in the background.",
      "The walls are lined with bulletin boards",
      "and posters filled with information and updates related to ongoing cases."
    ]
    [policeman]
    [ simpleItem "handcuffs",
      simpleItem "gun",
      simpleItem "ammunition",
      simpleItem "suitcase"
    ]
    ( Paths
        (Path "City Square")
        InvalidPath
        (Path "Policeman Office")
        (Path "Cage")
    )

policemanOffice =
  Location
    "Policeman Office"
    [ "You are in office of a policeman who is very busy is cluttered,",
      "with piles of paperwork and files,",
      "as well as various items of police equipment scattered throughout the room.",
      "The desk is usually covered with a computer,",
      "radio, phone, and other tools of the trade,",
      "with little space for anything else."
    ]
    [sheriff]
    [ simpleItem "pen",
      simpleItem "paper",
      simpleItem "blank_paper"
    ]
    ( Paths
        InvalidPath
        InvalidPath
        InvalidPath
        (Path "Police Station")
    )

-- TODO: weed if holding clover
cage =
  Location
    "Cage"
    [ "You see cells at the police station",
      "where there are many detainees are small and cramped,",
      "with just enough room",
      "for a narrow bed and a toilet in the corner.",
      "The walls are made of cold, gray concrete",
      "and the only light comes from a small window high up near the ceiling."
    ]
    [scaryPrisoner]
    [ makeshiftKnife,
      simpleItem "weed"
    ]
    ( Paths
        InvalidPath
        InvalidPath
        (Path "Police Station")
        InvalidPath
    )

policeLocations =
  [ policeStation,
    policemanOffice,
    cage
  ]